File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.py
n = int(input())
arr = list(map(int,input().split()))
stack2 = []
target = 1
while arr:
    if target == arr[0]:
        target += 1
        arr.pop(0)
    else:
        stack2.append(arr.pop(0))
    while stack2:
        if stack2[-1] == target:
            stack2.pop()
            target+=1
        else:
            break
if len(stack2) == 0:
    print('Nice')
else:
    print('Sad')

File: 백준/Silver/12789. 도키도키 간식드리미/도키도키 간식드리미.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int num = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        Queue<Integer> que = new LinkedList<>();
        Stack<Integer> stack = new Stack<>();

        for(int i =0;i<num;i++){
            que.add(Integer.parseInt(st.nextToken()));
        }

        int student = 1;
        while (!que.isEmpty() || !stack.isEmpty()){
            if(!que.isEmpty() && que.peek() == student){
                que.poll();
                student ++;
            } else if(!stack.isEmpty() && stack.peek() == student){
                stack.pop();
                student ++;
            }

            else if(!que.isEmpty()){
                stack.push(que.poll());
            } else {
                System.out.println("Sad");
                return;
            }
        }
        System.out.println("Nice");
    }
}
